### PR TITLE
Extend service API

### PR DESCRIPTION
**Pull request by imobachgs**
This PR extends `SystemdService` API to properly support sockets. Now, instead of using `enable/disable`, we can specify the `start_mode` which can be `:boot`, `:demand` and `:manual`. Additionally, the API supports querying which modes are supported for a given service.

### CODE REVIEW

File: library/systemd/src/modules/systemd_service.rb [hunk 1]
```diff
@@ -174,18 +174,118 @@ def restart
 
       # Returns socket associated with service or nil if there is no such socket
       def socket
+        return @socket if @socket
```
Comment (jreidinger) [inline library/systemd/src/modules/systemd_service.rb]: this will work only if socket is not found. But I expect that check for triggered_by is quick enough. What I do not like is that returned socket is cached for long time and it still has old status. Maybe at least refresh_properties should be called?

File: library/systemd/src/modules/systemd_service.rb [hunk 2]
```diff
@@ -174,18 +174,118 @@ def restart
...
+      def socket?
+        !socket.nil?
+      end
+
+      # Return the start mode
+      #
+      # @return [Symbol] Start mode (:boot, :demand, :manual)
+      def start_mode
+        return :boot if enabled_on_boot?
+        return :demand if enabled_on_demand?
```
Comment (jreidinger) [inline library/systemd/src/modules/systemd_service.rb]: similar issue I am solving in my class. What is start mode when service and also socket is enabled? I agree that boot makes sense, but we should document it.

File: library/systemd/src/modules/systemd_service.rb [hunk 3]
```diff
@@ -174,18 +174,118 @@ def restart
...
+        :manual
+      end
+
+      # Set the service start mode
+      #
+      # See {#start_modes} to find out the supported modes for a given service (usually :boot,
+      # :manual and, in some cases, :demand).
```
Comment (jreidinger) [inline library/systemd/src/modules/systemd_service.rb]: I do not see it documented in start_mode.

File: library/systemd/src/modules/systemd_service.rb [hunk 4]
```diff
@@ -174,18 +174,118 @@ def restart
...
+      #
+      # @see #start_modes
+      def start_mode=(mode)
+        if !start_modes.include?(mode)
+          log.warn "Invalid start mode: '#{mode}' for service '#{name}'"
```
Comment (jreidinger) [inline library/systemd/src/modules/systemd_service.rb]: agh...no, please no silent ignore of invalid value. Just use `ArgumentError` like any other nice method.

File: library/systemd/src/modules/systemd_service.rb (hunk 3 above)
Comment (jreidinger) [reply library/systemd/src/modules/systemd_service.rb]: ah, I see start_mode and not start_modes

File: library/systemd/src/modules/systemd_service.rb [hunk 5]
```diff
@@ -174,18 +174,118 @@ def restart
...
+
+      # Return the list of supported start modes
+      #
+      # * :boot:   The service will be started when the system boots.
+      # * :manual: The service is disabled and it will be started manually.
+      # * :demand: The service will be started on demand (using a Systemd socket).
+      #
+      # @return [Array<Symbol>] List of supported modes.
+      def start_modes
+        return @start_modes if @start_modes
```
Comment (jreidinger) [inline library/systemd/src/modules/systemd_service.rb]: I see here one risk. We use `find` for socket and not `find!` so when due to some temporary error it is not found or some systemctl failed, it will never can use on demand.

File: library/systemd/src/modules/systemd_service.rb (hunk 3 above)
Comment (imobachgs) [reply library/systemd/src/modules/systemd_service.rb]: `start_modes` (see the final 's'). However, I was thinking about calling it `supported_start_modes` or something like that to be more clear.

File: library/systemd/src/modules/systemd_service.rb (hunk 5 above)
Comment (jreidinger) [inline library/systemd/src/modules/systemd_service.rb]: I see here one risk. We use `find` for socket and not `find!` so when due to some temporary error it is not found or some systemctl failed, it will never can use on demand.

File: library/systemd/src/modules/systemd_service.rb (hunk 2 above)
Comment (imobachgs) [reply library/systemd/src/modules/systemd_service.rb]: Yes, it is `:boot`. I will document it.

File: library/systemd/src/modules/systemd_service.rb [hunk 6]
```diff
@@ -174,18 +174,118 @@ def restart
...
       end
 
     private
 
+      # Determine whether the service is enabled on demand
+      #
+      # @return [Boolean] true if it is enabled on demand; false otherwise.
+      def enabled_on_demand?
+        return false unless socket
+        socket.enabled?
```
Comment (jreidinger) [inline library/systemd/src/modules/systemd_service.rb]: why not `socket && socket.enabled?` and then separated method is probably not needed?

File: library/systemd/src/modules/systemd_service.rb (hunk 1 above)
Comment (imobachgs) [reply library/systemd/src/modules/systemd_service.rb]: I just wanted to avoid calling systemctl several times. For instance, if we do something like:

```ruby
service.socket.disable if service.socket?
```

We would be calling `#socket` twice. Having some kind of `reload` might help (which could be implemented calling `#find` again).

File: library/systemd/test/systemd_service_test.rb [hunk 7]
```diff
@@ -130,6 +130,201 @@ module Yast
...
+      subject(:service) { SystemdService.find("cups") }
+      let(:socket) { double("socket", disable: true) }
+
+      before do
+        stub_services(service: "cups")
+        allow(service).to receive(:socket).and_return(socket)
+        allow(service).to receive(:disable)
+      end
+
+      context "when no argument is given" do
```
Comment (jreidinger) [inline library/systemd/test/systemd_service_test.rb]: ugh, what? it does not reflect test.

File: library/systemd/src/modules/systemd_service.rb (hunk 5 above)
Comment (imobachgs) [reply library/systemd/src/modules/systemd_service.rb]: I just wanted to avoid asking `systemctl` more than necessary. IMHO, it looks really like a corner case. I would expect more problems if systemctl is failing (even temporarily).

File: library/systemd/test/systemd_service_test.rb [hunk 8]
```diff
@@ -130,6 +130,201 @@ module Yast
...
+    end
+
+    describe "#enable" do
+      subject(:service) { SystemdService.find("cups") }
+
+      before do
+        stub_services(service: "cups")
+      end
+
+      it "sets start_mode to :boot" do
```
Comment (jreidinger) [inline library/systemd/test/systemd_service_test.rb]: I am not sure if this is good test and documention. You want to be sure that when you call enable that it "enable service". This is important and not some internal method for start_mode.

File: library/systemd/test/systemd_service_test.rb (hunk 7 above)
Comment (imobachgs) [reply library/systemd/test/systemd_service_test.rb]: Oh, a leftover.

File: library/systemd/src/modules/systemd_service.rb [hunk 9]
```diff
@@ -174,18 +174,118 @@ def restart
...
+      alias_method :enabled_on_boot?, :enabled?
+      private :enabled_on_boot?
+
+      # Determine whether the service is enabled or not
+      #
+      # The service can be enable to be started on boot or on demand.
+      #
+      # @return [Boolean] true if the service is enabled; false otherwise.
+      def enabled?
+        start_mode != :manual
```
Comment (joseivanlopez) [inline library/systemd/src/modules/systemd_service.rb]: I don't understand this method. `enabled?` definition is based on `start_mode`, but `start_mode` definition is also based on `enabled?`. Probably I am missing something.

File: library/systemd/test/systemd_service_test.rb [hunk 10]
```diff
@@ -130,6 +130,201 @@ module Yast
...
+    end
+
+    describe "#disable" do
+      subject(:service) { SystemdService.find("cups") }
+
+      before do
+        stub_services(service: "cups")
+      end
+
+      it "sets start_mode to :manual" do
```
Comment (jreidinger) [inline library/systemd/test/systemd_service_test.rb]: same here

File: library/systemd/test/systemd_service_test.rb (hunk 8 above)
Comment (imobachgs) [reply library/systemd/test/systemd_service_test.rb]: I wanted to avoid repetition. To be honest, `enable` now sets the start mode to `:boot` (that's what it does).

File: library/systemd/src/modules/systemd_service.rb (hunk 2 above)
Comment (joseivanlopez) [reply library/systemd/src/modules/systemd_service.rb]: IMHO, start mode should be `:demand` as long as the socket is enabled. In fact, whether the service is enabled or not should not be relevant in that case.

File: library/systemd/src/modules/systemd_service.rb (hunk 1 above)
Comment (joseivanlopez) [reply library/systemd/src/modules/systemd_service.rb]: If socket is not found in a first call, it will try to find it again in subsequent calls. Does it make sense? I don't know.

File: library/systemd/test/systemd_service_test.rb (hunk 8 above)
Comment (jreidinger) [reply library/systemd/test/systemd_service_test.rb]: well, but we have this method just to be backward compatible. So we should be sure it do what it did before. So we do not break other modules.

File: library/systemd/src/modules/systemd_service.rb [hunk 11]
```diff
@@ -230,10 +230,16 @@ def socket?
 
       # Return the start mode
       #
+      # See {#start_modes} to find out the supported modes for a given service (usually :boot,
+      # :manual and, in some cases, :demand).
+      #
+      # When the service are the service (:boot) and the socket (:demand) are enabled,
```
Comment (jreidinger) [inline library/systemd/src/modules/systemd_service.rb]: double "are"

File: library/systemd/src/modules/systemd_service.rb (hunk 9 above)
Comment (imobachgs) [reply library/systemd/src/modules/systemd_service.rb]: No, `start_mode` is not based on `enabled?`. Take into account that `#socket` returns an instance of a different class.

File: library/systemd/test/systemd_service_test.rb [hunk 12]
```diff
@@ -276,26 +276,32 @@ module Yast
...
       before do
         stub_services(service: "cups")
+        allow(service).to receive(:socket).and_return(socket)
       end
 
-      it "sets start_mode to :boot" do
-        expect(service).to receive(:start_mode=).with(:boot)
+      it "enables the service to start on boot" do
+        expect(service).to receive(:enable_service)
+        expect(socket).to receive(:disable)
```
Comment (jreidinger) [inline library/systemd/test/systemd_service_test.rb]: hmm, that in fact realize me that we are not 100% backward compatible, so maybe create separate `it` state for this side effect. As previously service.enable does not disable socket. So we should document it.

File: library/systemd/src/modules/systemd_service.rb (hunk 9 above)
Comment (imobachgs) [reply library/systemd/src/modules/systemd_service.rb]: Clarified offline. `enabled?` (from parent class) is now `enabled_on_boot?` and `enabled?` has been redefined.

Comment (jreidinger) [reply library/systemd/src/modules/systemd_service.rb]: this is also backward incompatible change, but I think it is fine.
